Lib/test/test_doctest.py

```diff
@@ -2659,12 +2662,42 @@ def test_testfile(): r"""
     >>> sys.argv = save_argv
 """
 
+class TestLoader(importlib.abc.MetaPathFinder, importlib.abc.ResourceLoader):
+
+    def find_spec(self, fullname, path, target=None):
+        return importlib.util.spec_from_file_location(fullname, path, loader=self)
+
+    def get_data(self, path):
+        with open(path, mode='rb') as f:
+            return f.read()
+
+class TestHook:
+
+    def __init__(self, pathdir):
+        self.sys_path = sys.path[:]
+        self.meta_path = sys.meta_path[:]
+        self.path_hooks = sys.path_hooks[:]
+        sys.path.append(pathdir)
+        sys.path_importer_cache.clear()
+        self.modules_before = sys.modules.copy()
+        self.importer = TestLoader()
+        sys.meta_path.append(self.importer)
+
+    def remove(self):
+        sys.path[:] = self.sys_path
+        sys.meta_path[:] = self.meta_path
+        sys.path_hooks[:] = self.path_hooks
+        sys.path_importer_cache.clear()
+        sys.modules.clear()
+        sys.modules.update(self.modules_before)
```

Comment:
It would be better to do this in a context manager just to be very sure we don't miss the cleanup.  There are some helpers in `test.test_importlib.util` that may be helpful here, though I confess I don't understand the inner workings of `importlib` well enough to know exactly what we need here :).  On the other hand, perhaps we can just pass in a mock object for `package` and avoid messing with global import state at all.

Reply:
I looked through `test.test_importlib.util` and didn't see anything useful. I'm also not confident enough in my knowledge of how packages work to try to construct a mock package object that won't cause some other problem.

The context manager suggestion is a good one. It looks straightforward enough, I will test that and push an update if I can get it working.

Reply:
The context manager turned out to be very simple; `contextlib.contextmanager` is my friend. :-)

There are a total of three new commits:

Commit efabd42 does the breakup of the long line (I just committed your suggested change, zware).

Commit 1d56c05 fixes a typo in the breakup of the long line.

Commit e1f0a2f implements the context manager.

Reply:
@zware Thanks for the kudos, btw!

Reply:
You want `test.test_importlib.util.import_state()`: https://github.com/python/cpython/blob/master/Lib/test/test_importlib/util.py#L183

Reply:
@pdonis, did you see @brettcannon's message here?  I think there is some opportunity to clean up the clean up code a bit, but on the other hand that can be done in a follow-on PR.